MapboxCoreNavigationTests/RouteControllerTests.swift

```diff
@@ -113,4 +134,69 @@ class RouteControllerTests: XCTestCase {
     }
+
+    func testUserWithinTunnelEntranceRadius() {
+        let navigation = tunnelSetup.routeController
+        
+        // Intersection with a tunnel roadClass
+        let intersectionLocation = navigation.routeProgress.route.legs[0].steps[1].intersections![1].location
+        let tunnelEntranceLocation = CLLocation(latitude: intersectionLocation.latitude, longitude: intersectionLocation.longitude)
+        
+        navigation.locationManager(navigation.locationManager, didUpdateLocations: [tunnelEntranceLocation])
+
+        let currentLegProgress = (navigation.routeProgress.currentLegProgress)!
+        let currentStep = currentLegProgress.currentStep
+        let currentIntersection = currentStep.intersections![currentLegProgress.stepIndex]
+        
+        var currentLocation = location(at: navigation.location!.coordinate)
+        var distanceToTunnelEntrance = currentIntersection.location.distance(to: currentLocation.coordinate)
+        
+        var userIsAtTunnelEntranceRadius = navigation.userWithinTunnelEntranceRadius(at: currentLocation, intersection: currentIntersection, distance: distanceToTunnelEntrance)
+        XCTAssertTrue(userIsAtTunnelEntranceRadius, "Location must be within the tunnel entrance radius")
+
+        let outsideTunnelEntranceRadius =  intersectionLocation.coordinate(at: 200, facing: intersectionLocation.direction(to: tunnelSetup.firstLocation.coordinate))
+        let outsideTunnelEntranceRadiusLocation = CLLocation(latitude: outsideTunnelEntranceRadius.latitude, longitude: outsideTunnelEntranceRadius.longitude)
+        
+        navigation.locationManager(navigation.locationManager, didUpdateLocations: [outsideTunnelEntranceRadiusLocation])
+        
+        currentLocation = location(at: navigation.location!.coordinate)
+        distanceToTunnelEntrance = currentIntersection.location.distance(to: outsideTunnelEntranceRadiusLocation.coordinate)
+
+        userIsAtTunnelEntranceRadius = navigation.userWithinTunnelEntranceRadius(at: currentLocation, intersection: currentIntersection, distance: distanceToTunnelEntrance)
+        XCTAssertFalse(userIsAtTunnelEntranceRadius, "Location must not outside the tunnel entrance radius")
+    }
+    
+    fileprivate func location(at coordinate: CLLocationCoordinate2D) -> CLLocation {
+        return CLLocation(coordinate: coordinate,
+                            altitude: 5,
+                  horizontalAccuracy: 258.20,
+                    verticalAccuracy: 200,
+                              course: 20,
+                               speed: 15,
+                           timestamp: Date())
+    }
+    
+    // TODO: Test Tunnel Animation Enabled
+    // TODO: Test Tunnel Animation Disabled
+    func testTunnelSimulatedNavigation() {
+        
+    }
+    
+    func testEnableTunnelAnimation() {
+        let navigation = tunnelSetup.routeController
+        
+        // Intersection with a tunnel roadClass
+        navigation.advanceStepIndex(to: 1)
```

Comment:
We should not be advancing the step index with this method. By updating the index with `locationManager(_:didUpdateLocations:)`, we can verify all pieces of the puzzle are working properly.

Reply:
I'd be curious to understand why not? There's a place and time to test things working properly together (an integrated test or collaboration test), but there also is value in testing directly at a different seam in the object graph especially when a pleasant higher-level API is available or aspects of the implementation have undesirable side effects.

In this case the SUT has a call to performSelector(_ withDelay:) which is missing a testing shim. Therefore we need to be careful to dance around it a bit under test. Not sure if that's a factor here or not.

We should be careful not to always reach for the lowest-level way of doing things when a higher-level abstraction exists and is appropriate to use. Is it not appropriate in this case or are you more worried about not having integrated test coverage?